Expose YUV Comparison and Utilities

ForestedDev opened the pull request:
To solve #12 in a fairly user friendly way, exposes the `Decompose` trait, rgb/yuv conversion, and a `yuv_hybrid_compare` method. As well as fixing some visibility issues that arose from exposing the `utils` module.

These changes are hidden behind the "yuv_compare" feature flag since it will, in general, probably *not* be needed by the majority of users. But in more performance critical areas it can yield a good speed boost such as in [my more recent project](https://codeberg.org/vortetty/SVGIZE) which benefited greatly from swapping to yuv to avoid a significant amount of data copying and format conversion. In other cases where one can use yuv rather than rgb, a similar potential exists. The speed boost is not as great as one may see with gpu-based comparison however it is enough to justify the possibility, saving 5-10% of runtime for me in some cases.

Commit:
add yuv_hybrid_compare and "advanced_functions"

Commit:
Fixes

Commit:
Fix compilation

Commit:
Fix naming, and visibility

Commit:
Fix bug in yuv_hybrid_compare, reduce redundant checks
The internal_yuv_hybrid_compare function was doing the image size check which doubles the checks over internal_yuv_hybrid_compare's. move this to yuv_hybrid_compare and check all 3 channels to avoid double checking when using rgb images and check all channels with yuv_hybrid_compare

Commit:
Add docs

ChrisRega commented on src/hybrid.rs:
```diff
@@ -179,6 +179,35 @@ pub fn rgb_hybrid_compare(first: &RgbImage, second: &RgbImage) -> Result<Similar
+/// and and the u and v color diffs leading to color-deviations in green, blue and cyan
+/// All-black meaning no differences
+#[cfg(feature = "yuv_compare")]
+pub fn yuv_hybrid_compare(first_channels: &[GrayImage; 3], second_channels: &[GrayImage; 3]) -> Result<Similarity, CompareError> {
```
Could you please add a unit test covering this?

ForestedDev replied:
tests added in latest commit

ChrisRega commented on src/lib.rs:
```diff
@@ -77,8 +77,14 @@ mod histogram;
 mod hybrid;
 mod squared_error;
 mod ssim;
+
+#[cfg(not(feature="yuv_compare"))]
 mod utils;
 
+/// Provides some utilities to make yuv image management and conversion easier.
+#[cfg(feature="yuv_compare")] // Exposes rgb/yuv conversions and split to yuv publicly, others were left to be pub crate or private
+pub mod utils;
+
```
This behavior is completely new and should be covered by tests just to pin the api, as we do with all public API.

ForestedDev replied:
The pub/non-pub cannot be tested since there isn't really a way to check if it is public or not during a test, but i'll get tests added covering the exposed APIs

ForestedDev replied:
`trait Decompose` is now covered by tests, and `rgb_to_yuv` and `yuv_to_rgb` are covered by preexisting tests, and their publicity is also now explicitly tested for. These are currently the only 3 fully public APIs exposed here, the rest are crate-level publicity or private to the `utils` module. Compilation will fail should publicity be broken for these.

This is the closest that can be gotten for a test, and works since the cucumber test framework scripts can only access `mod utils` if the `yuv_compare` feature is enabled. There is sadly not a way to guarantee utils is private when it's disabled that i can find.

ChrisRega commented on src/utils.rs:
```diff
@@ -113,7 +116,7 @@ impl Decompose for RgbImage {
 }
 
-pub fn merge_similarity_channels(input: &[&GraySimilarityImage; 3]) -> RGBSimilarityImage {
+pub(crate) fn merge_similarity_channels(input: &[&GraySimilarityImage; 3]) -> RGBSimilarityImage {
```
thanks for the cleanup :)

ForestedDev replied:
Of course, exposing utils as public added irrelevant things to public, so it seemed like a good thing to fix :)

ChrisRega reviewed (changes requested):
Thank you for the suggestion, it looks really good and helpful to me, especially since it was spawned by a user request. Please add feature-toggle guarded unit tests to pin the API behavior as we dont ever want to break public API without knowing.

Commit:
yuv_hybrid_compare and decompose tests

Commit:
Add test for publicity on yuv/rgb conversion functions

Commit:
Cleanup: avoid hiding lifetime that's elided elsewhere
avoids using mixed syntax for lifetimes, per clippy's #[warn(mismatched_lifetime_syntaxes)]

ChrisRega commented on tests/features/decompose.feature:
```diff
@@ -0,0 +1,35 @@
+Scenario Outline: Compare newly split yuv channels to expected values
+  Given the image '<full_image>' is loaded
+  When comparing yuv channels
+  Then check split success
```
```suggestion
  Then the split channels match the references
```

then steps should be formulated in passive, but this is just  a hint - I can fix this later :)

ForestedDev replied:
i will keep that in mind for future use- was the first time i'd worked with that test runner

ChrisRega commented:
Will merge tomorrow and draft release, thank you for the contribution :)
